### PR TITLE
Error Bar Resurrection

### PR DESCRIPTION
**Pull request by M-Kusumgar**
## Description
This PR aims to add error bars back into Hint. Unfortunately, there is no vue 3 version of chart js error bars plugin so I am using the annotation plugin to draw the 3 lines for an error bar (the two horizontal lines and one vertical line). I have also included tests for the vue-charts folder. It makes sense to leave these components in Hint because Hint is the only thing that uses vue-charts.

As a side note, you can click the legend to hide bars in chart js and the production version of hint actually doesn't work with this feature (you can try it out on the production version, the error bars go weird)

### CODE REVIEW

**Review by EmmaLRussell — COMMENTED** [review]
Looks great!

The error values were previously included in the tooltips, but this no longer seems to be the case. Separate ticket?

If we're making the decision to keep Vue3 barchart solely within HINT, then the barchart should use the existing FilterSelect, in components/plots, and all its source should move out of `vue-chart`.. or is there a ticket or plan to do that in a separate branch?

All the stuff working around the animation seems like a faff and potentially quite fragile. Are the animations actually useful? I guess we could just turn them off (though they are pretty nifty!)

File: src/app/static/src/app/components/plots/utils.ts [hunk 1]
```diff
@@ -250,7 +250,7 @@ export const flattenXAxisFilterOptionIds = (selections: BarchartSelections, filt
 
 
 export const updateSelectionsAndXAxisOrder = (data: BarchartSelections, selections: BarchartSelections, flattenedXAxisFilterOptionIds: string[], updateSelections: (data: {payload: BarchartSelections}) => void) => {
-    const payload = {...selections, ...data}
+    const payload = {...selections, ...structuredClone(data)}
```
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply src/app/static/src/app/components/plots/utils.ts:253]: apparently it is not necessary now? I was getting an error about mutating data when it was a const or something like that but I just reverted it and there are not errors? have removed it!

File: src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue [hunk 2]
```diff
@@ -111,118 +94,180 @@ export default defineComponent({
...
+                    (out: number[], bool: boolean, index: number) => bool ? out.concat(index) : out, 
+                    []
+                );
+
+                const numOfDatasets = this.showLabelErrorBars.filter(x => x).length;
+                const numOfLabels = chartData.labels.length;
+                const numOfBars = numOfDatasets * numOfLabels;
+
+                // amount of padding chart js uses by default for each bar 
+                const barPercentage = 0.8;
```
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue:115]: done

File: src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue [hunk 3]
```diff
@@ -111,118 +94,180 @@ export default defineComponent({
...
+                            So when there are a few wide bars, numOfBars/(numOfBars + 10) is small
+                            which solves our problem. When there are a lot of thin bars
+                            numOfBars/(numOfBars + 10) is closer to 1 so we don't get very small
+                            error bar widths
+                        */
+                        const errorBarWidth = (numOfBars/(numOfBars + 10)) * halfBarWidth * 0.3;
+
+                        if (errorBarData[xLabel].minus && errorBarData[xLabel].plus) {
+                            const config = [
+                                {
```
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue:147]: done

File: src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue [hunk 4]
```diff
@@ -111,118 +94,180 @@ export default defineComponent({
...
+                    ci.hide(index);
+                }, 50)
+                legendItem.hidden = true;
+            } else {
+                this.showLabelErrorBars[index] = true;
+                setTimeout(() => {
+                    ci.show(index);
+                }, 50)
+                legendItem.hidden = false;
+            }
```
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue:247]: i think this is cleaner for sure, now you would think this is equivalent to what I wrote right? somehow it is not, chart js really doesnt like the ternary, and threw an error so have had to write two timeouts but I have condensed the other stuff!

File: src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue [hunk 5]
```diff
@@ -111,118 +94,180 @@ export default defineComponent({
...
+    },
+    methods: {
+        hideErrorBarsDuringAnimation() {
+            this.displayErrorBars = false;
+            if (this.errorBarTimeout) {
+                window.clearTimeout(this.errorBarTimeout)
+            }
+            this.errorBarTimeout = setTimeout(() => {
+                this.displayErrorBars = true
+            }, 1050)
```
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue:224]: yep its a bit nicer without the timeout actually! managed to hook it up

**Comment by M-Kusumgar** [comment]
> The error values were previously included in the tooltips, but this no longer seems to be the case. Separate ticket?

Actually this was a good catch, it seems chart js changed where the tooltip callbacks should be placed in the options and also changed what gets passed into the tooltip callback functions! All fixed now!

> If we're making the decision to keep Vue3 barchart solely within HINT, then the barchart should use the existing FilterSelect, in components/plots, and all its source should move out of vue-chart.. or is there a ticket or plan to do that in a separate branch?

Have made a ticket about moving the files: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4526/Move-vue-charts-folder-components-into-main-src-folder, the filterSelects however are different! Shall I make a ticket to merge them? It might make the component a bit complicated though

> All the stuff working around the animation seems like a faff and potentially quite fragile. Are the animations actually useful? I guess we could just turn them off (though they are pretty nifty!)

Jeff and rachel did seem to like them, I think they are neat. I would like to retain animations if possible, I also think they are less fragile with the onComplete animation event!

File: src/app/static/src/app/components/plots/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply src/app/static/src/app/components/plots/utils.ts:253]: Seems odd but ok!

**Review by EmmaLRussell — COMMENTED** [review]
Tooltips looking great for proportions, but there's an issue with numeric values (this is PLHIV) :

Screenshot from 2023-08-16 10-12-15

> the filterSelects however are different!

Oh no! Sounds like they could probably be rationalised somehow - I guess they must share some code? Might not be combined into one, but could at least share code from one place, and should probably have different names. 

> chart js really doesnt like the ternary

I don't understand that though, because it's not chartjs that's dealing with the ternary, but the component code. It should just be choosing which function to call. But anyway, looks a lot terser now so 🤷 

Yeah, it's definitely nicer hooking into onComplete, let's keep the flashy animations! 😆

**Review by LekanAnni — APPROVED** [review]
LGTM 👍

**Comment by M-Kusumgar** [comment]
> Oh no! Sounds like they could probably be rationalised somehow - I guess they must share some code? Might not be combined into one, but could at least share code from one place, and should probably have different names.

they do share a decent ish amount of code, maybe in that ticket for moving the components to src from vue-charts we can also add merge the filter selects? because there is already something in that ticket that says merge utils

> I don't understand that though, because it's not chartjs that's dealing with the ternary, but the component code. It should just be choosing which function to call. But anyway, looks a lot terser now so 🤷

literally no clue, was so odd honestly

> Tooltips looking great for proportions, but there's an issue with numeric values (this is PLHIV) :

realised what the problem was! was looking at the wrong property of the context object passed into the tooltip callback, should be all good now!

**Review by EmmaLRussell — APPROVED** [review]
:+1: 

> maybe in that ticket for moving the components to src from vue-charts we can also add merge the filter selects? because there is already something in that ticket that says merge utils

sounds good to me!

File: src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue [hunk 6]
```diff
@@ -71,9 +71,10 @@ export default defineComponent({
                                 const {
                                     datasetIndex,
                                     dataset,
-                                    formattedValue: yLabel,
+                                    raw: yLabel,
                                     label: xLabel
                                 } = context;
+                                console.log(context)
```
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply src/app/static/src/app/vue-chart/src/bar/BarChartWithErrors.vue]: missed this! removed!

**Pull request merged**